tree/scenario/Scenario_manager.py

```diff
@@ -0,0 +1,159 @@
+        if not(self.scenario_select):
+            # manager not initialize
+            # i could append with linked scenarios
+            self.mutex.release()
+            return 
+        if scenar.get_marker() == MARKER.NONE:
+            # none scenario are just invisible
+            scenar.do()
+            self.mutex.release()
+            return
+        elif scenar.get_marker() != MARKER.OFF:
+            # if it is ON
+            self.do(scenar)
+        elif self.scenario_select.get_marker() == MARKER.OFF:
+            # if we are already OFF, so just clear the list an shutdown 
+            # TODO strange comportements
+            self.do(scenar)
+            self.stack.clear()
+        else: 
+            if len(self.stack) == 0:
+                self.do(scenar)
+            else:
+                # do the top of the stack if it is not empty
+                self.do(self.top())
+        self.scenario_select = scenar
+        self.mutex.release()
+
+    def do_scenar_secondary(self, scenar):
+        """
+        This method is call to modifie secondaries scenarios to and in the stack
+        """
+        assert self.scenario_select, "Need to initialize before"
+        self.mutex.acquire()
+        if scenar.get_marker() == MARKER.NONE:
+            # null scenario are just invisible
+            scenar.do()
+            self.mutex.release()
+            return
+        self.stack.append(scenar)
+        if self.scenario_select.get_marker() == MARKER.OFF:
+            self.do(scenar)
+        self.mutex.release()
+
+    def get_scenar_select(self):
+        return self.scenario_select
+
+    def get_current_scenar(self):
+        return self.current_scenar
+
+    def get_stack(self):
+        return self.stack
+
+    def reload_scenar_selected(self, scenario_select):
+        self.scenario_select = scenario_select
+
+    def reload_current_scenar(self, scenario):
+        self.current_scenar = scenario
+        self.current_scenar.do()
+
+    def reset(self):
+        self.stack.clear()
+        self.current_scenar.set_state(False)
+        self.current_scenar = None
+        self.scenario_select = None
+
+    def push(self, scenar):
+        self.stack.append(scenar)
+
+    def remove(self, scenar):
+        """
+        Call to remove a scenario for the stack
+        """
+        self.mutex.acquire()
+        if scenar.get_marker() == MARKER.NONE:
+            # none scenario are just invisible
+            self.mutex.release()
+            return
+        try:
+            self.stack.remove(scenar)
+        except ValueError:
+            pass
+        if self.current_scenar is scenar: # it is the current_scenar
+            # do the appropriate scenario
+            if len(self.stack) == 0 or self.scenario_select.get_marker() != MARKER.OFF:
+                self.do(self.scenario_select)
+            else:
+                self.do(self.top())
+        self.mutex.release()
+
+    def top(self):
+        return self.stack[-1]
```

Comment:
Top of the stack, because it's obvious when you don't read the source code that Scenario_manager has a stack